README.md
# RyToast Planner - A Toastmasters Meeting Role Scheduling Webapp
Rytoast Planner is a web app created to schedule club members for roles at the Ryerson Toastmasters. Created using Django 3.1.2, Python 3.9, Bootstrap 4.4.1, and Heroku Postgres.

This app was originally created as the capstone project for [CS50’s Web Programming with Python and JavaScript](https://cs50.harvard.edu/web/2020/) course.

Visit the live website: https://rytoastplanner.herokuapp.com/

![Meeting Page Screenshot](/meeting_page.PNG)



<br>

## Table of contents
---
- [Features](#features)
- [Files](#files)
- [Status](#status)
- [Inspiration](#inspiration)
- [Contact Information](#contact-information)



<br>

## Features
---
#### Meeting Management
- Executives can create, edit, and delete meetings.
- Set and manage the meeting start time, theme, and word of the day
- Manage the events that occur at each meeting (duration, description, role, order, user)
- View the meetings individually, as a list, or as a spreadsheet.
- Bulk create a group of meeting using the 'Bulk Create Meetings' button
- Send emails to members regarding attendance, role sign ups, and general meeting reminders

#### Individual Meeting View
- View all the events of the meeting, including start times, duration, member assigned, and event description
- A sidebar with meeting information, helpful links, a list of the club executives with links to their LinkedIn, a list of all the meeting roles with the members who are scheduled for them, and a guest list.
- Ability to confirm or deny attendance at the meeting
- Ability to print out a formatted meeting agenda

#### Meeting List View
- View the date of each meeting with a link to view the full meeting page
- View the number of speeches scheduled for each meeting
- View the theme for the meeting
- View the currently scheduled Toastmaster, General Evaluator, and Topicsmaster for each meeting

#### Spreadsheet View
- View all of the role holders for all of the future meetings in a spreadsheet format
- Members can sign up for any role at any meeting
- Executives can assign any available member to any future meeting
- Mark if the user is going to be attending the meeting or not

#### Role Scheduling
- Non-logged-in users can view all role holders for each meeting.
- Logged-In users can sign themselves up for roles at any meeting.
- Logged-In users can confirm or deny their attendance at a meeting. Denying it will remove them from the recommendation lists executives get while assigning members to roles.
- Executives can assign any member to any role for any meeting.
- Executives can remove members from their scheduled roles.
- Executives can see an ordered list of recommendations for which members should be scheduled for each role.

#### Authentication and Administration
- Register an account, log in, logout.
- Reset password via email confirmation
- Executives can create new member accounts.
- Executives are assigned using the Django Admin panel.
- Report a bug with the floating "Report a Bug" button.
- Website admin can view a sorted bug list on the Bug List page.



<br>

## Files
---
The following is a list of all the significant files in the web app and a description of their importance.
- The app uses an SQL database. In development, it is SQLite. In production, it is Heroku Postgres.
- **\RolesScheduler** contains the main Django section and settings.
    - **\RolesScheduler\urls.py** contains the redirects for the admin, authentication, and agenda app views.
- **\template\registration** contains all the HTML template files for the login and password reset pages.
- **\agenda** is the folder for the Django app.
    - **\agenda\templates\agenda** contains the HTML templates for the layout and the various meeting views.
    - **\agenda\static\agenda** contains the styles.css file with the custom CSS styles for the entire web app.
    - **\agenda\migrations** contains the migration information to create an SQL database for the web app.
    - **\agenda\templates\admin.py** allows the databases to be viewed from the Django admin panel.
    - **\agenda\templates\models.py** contains the models that the SQL database uses
        - There are models for the users, meetings, role lists, attendees, eventlists, and the bug list.
        - There is a form for the bug list.
        - There is also a function to display 'deleted' where a user used to be if their account is deleted.
    - **\agenda\templates\views.py** contains all the views and many of the functions of the web app.
        - The views are for the index, meeting, meeting list, edit meeting, create a meeting, delete a meeting, spreadsheet, report a bug, see the bug list, register an account, create a user, and send email confirmations.
        - The API functions it contains are to update attendance records, sign up for a role, make a recommendations list for one role, change an event number, and add a new meeting event.
    - **\agenda\templates\functions.py** contains all the helper functions for the web app. The functions are to:
        - Update a meeting role list
        - Update one role in the database
        - Create a meeting roles list
        - Create a new set of Eventlist models based on default meeting events
        - Make a role recommendation list
        - Sort a list of members
        - Get a list of past role holders for a specific role and meeting
        - Convert a display text role to the backend role shorthand.
    - **\agenda\templates\urls.py** contains all the URL patterns for the agenda app including:
        - spreadsheet, meeting, edit meeting, create a meeting, bulk create meetings, delete a meeting, meeting list, report bug, bug list, update a role description, register, and create a user.
        - It also includes URLs for the API routes to update attendance records, sign up for a role, make a recommendation list for a role, change an event number, and add a new event.



<br>

## Status
---
The web app is currently in a live alpha hosted using Heroku. It is still in active development and many features and bugs are left to add and fix.

If you are interested in attending a club meeting you can check out the [Ryerson Toastmasters Club Website](https://www.ryersontoastmasters.ca/) and sign up to attend a meeting on the [Eventbrite Page](https://www.eventbrite.ca/e/the-ryerson-toastmasters-weekly-membership-meeting-tickets-110513236064).

View the [live website](https://rytoastplanner.herokuapp.com/)

View the [CHANGELOG](CHANGELOG.md)

Future Features:
- Calendar page to see all meetings at a glance and add them to the user's 3rd party calendar.
- A panel for executives to manage members' accounts.
- Printer-friendly spreadsheet view
- Ability to create and manage new Toastmasters Clubs
- Ability for a user to manage their profile (Username, email, password, image, etc)



<br>

## Inspiration
---
This app was created as the capstone project for [CS50’s Web Programming with Python and JavaScript](https://cs50.harvard.edu/web/2020/) course. As such, it draws on the lessons and techniques taught in the course regarding Python, Django, SQL, and Javascript.

The front end of the web app was inspired by the [Toastmasters International Website](https://www.toastmasters.org/), the [Monday.com Dashboard](https://www.monday.com), and the [Dashboard Adobe XD UI kit](https://www.adobe.com/products/xd/features/ui-kits.html).

The [Python Django Tutorial: Deploying Your Application (Option #2) - Deploy using Heroku](https://youtu.be/6DI_7Zja8Zc) tutorial by Corey Schafer helped transfer the web app to Heroku.



<br>

## Contact Information
---
Name: Aidan Dennehy

Website/Portfolio: https://www.aidandennehy.ca

LinkedIn: https://www.linkedin.com/in/adennehy

Github: https://github.com/Dracfo

CHANGELOG.md
# CHANGELOG
All notable changes to this project will be documented in this file.

The format is adapted from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Unreleased]
---
### Added:
-

### Changed:
-

### Removed:
-

### Fixed:
-

### Security
-



## [1.1.0-beta] - 2021-Jan-20
---
### Added:
- Email dashboard to send a final reminder about the meeting to everyone attending
- Email dashboard to ask everyone with confirmed or unknown attendance who doesn't have a role to sing up for one
- New features to the README.md file

### Changed:
- The meeting editing permissions so that the Zoom Master and chair can assign memebrs to roles and send meeting emails

### Removed:
- Roles confirmation email dashboard from confirmation_emails.html because there is no function to confirm a role yet

### Fixed:
- 'Confirm NOT Attending' button on individual meeting page not responding

### Security
- Added @login_required to several views to prevent non-logged in users from accessign secure sections of the website



## [1.0.2-beta] - 2021-Jan-19
---
### Added:
- Meeting Theme to the spreadsheet view



## [1.0.1-beta] - 2021-Jan-19
---
### Changed:
- Spreadsheet view can only be accessed by loged in users
- Spreadsheet view only gets role recommendations if the user is an executive



## [1.0.0-beta] - 2021-Jan-19
---
### Added:
- Print button to individual meeting page
- Resources navbar dropdown with links to the club role descriptions and basecamp
- An editable guest list to the individual meeting page
- Ability to mark yourself as absent or attending from the spreadsheet view

### Changed:
- on_delete value of Foreign Keys in the rolelist model to be SET_NULL so that all the meeting rolelists aren't deleted when a user is deleted.

### Removed:
- Ability to create an account, it now has to be done by an executive.



## [0.7.1-alpha] - 2021-Jan-18
---
### Security
- Set CSRF_COOKIE_SECURE = False and SESSION_COOKIE_SECURE = False to try and fix issues with logging in.



## [0.7.0-alpha] - 2021-Jan-18
---
### Added:
- Latest features to the README.md file
- Zoom Master role
- An email confirmation view for every meeting which can be used to send confirmation emails
- Ability for members to confirm their attendance at a meeting through a url sent to their email


### Removed:
- Removed attendance information and bug button from the page when it is printed
- Removed Facilitator and Ballot Counter from displaying in individual meeting and spreadsheet pages



## [0.6.2-alpha] - 2021-Jan-16
---
### Fixed:
- Spreadsheet now displays meetings ordered by date instead of meeting number




## [0.6.1-alpha] - 2021-Jan-16
---
### Added:
- Ability for members to sign up and reject roles in the spreadsheet view using the 'Sign Up' and 'X' buttons

### Fixed:
- 'Sign Up' button not working on the individual meeting page
- Member not being able to remove themselves from a role on individual meeting page



## [0.6.0-alpha] - 2021-Jan-15
---
### Added:
- Meeting themes to the meeting list view
- Bulk Create Meeting button to the navbar in the executive dropdown
- Added the ability to create meetings in bulk with the Bulk_Create_Meetings view

### Changed:
- Edit_meeting view now redirects you to the page for the meeting you just edited instead of the index page

### Fixed:
- Removed 'Edit Meeting' button for non-executives on individual meeting pages
- Removed 'X' button from the Absences and Meeting_id row in the spreadsheet view
- Removed the Meeting_id row from the spreadsheet view



## [0.5.0-alpha] - 2021-Jan-14
---
### Changed:
- Default description for the Chairperson's introduction
- Moved update_one_role_in_database, create_default_eventlist, role_recommendation_list, past_role_holders, and convert_role_to_shorthandfuncions to functions.py from views.py
- Changed redirection when there isn't an upcoming meeting to the meeting list. Used to be the login apge, which was confusing.

### Removed:
- Padding from the description textarea in the edit_meeting view

### Security
- Changed DEBUG setting value to an environment variable so that it is true locally and false in production
- Changed email password to be inputted using an environment variable
- Also changed the gmail password so that past git commits would not compromise future security of the email account
- Changed CSRF_COOKIE_SECURE and SESSION_COOKIE_SECURE to be True



## [0.4.1-alpha] - 2021-Jan-11
---
### Fixed:
- Error with creating a meeting due to a 20 character limit on role names being hit by the Table Topics Evaluator. Changed to 30 character limit.



## [0.4.0-alpha] - 2021-Jan-11
---
### Added:
- Links to role descriptions on the individual page sidebar
- Ability to edit meeting events in the edit_meeting view
- Ability for memebrs to update the speech description in the individual meeting page if they are signed in and signed up for that role

### Changed:
- Individual meeting page to render using the Eventlist database instead of the hardcoded string, sets up the eventlist to become editable in future.
- Schedule view now doesn't show past meetings
- Moved update_meeting function to the new functions.py file for non-view functions
- Speech descriptions to include asterisks at beginning and end to differentiate the title from the description

### Fixed:
- Name appearing twice on individual meeting page
- 'X' button not working in eventlist of individual meeting page
- Homepage not directing to the next meeting from today
- Meeting List View listing meetings in reverse chronological order



## [0.3.1-alpha] - 2021-Jan-07
---
### Added:
- Login page now displays errors if login information is incorrect or entered incorrectly

### Removed:
- Sergeant at arms from meeeting templates

### Fixed:
- Meeting Word of the Day and Theme not updating through the edit meeting form

agenda/functions.py
from django.db.models import Q
from datetime import datetime, timedelta, time, date

from .models import User, Meeting, Rolelist, Attendee, BugForm, Buglist, Eventlist


def update_meeting(meeting, new_roles):

    print(f'new_roles: {new_roles}')

    # Update the meeting time
    new_datetime = new_roles['date'] + " " + new_roles['start_time']
    meeting.starttime = new_datetime

    # Update the meeting WoD and Theme
    meeting.wod = new_roles['wod']
    meeting.theme = new_roles['theme']

    meeting.save()

    # Get the meeting information
    roles = Rolelist.objects.get(meeting=meeting)

    # Create a list of all the old roles and who has filled them
    rolelist = {
        'Toastmaster': roles.toastmaster,
        'Chair': roles.chair,
        'Facilitator': roles.facilitator,
        'Zoom Master': roles.zoom,
        'General Evaluator': roles.geneval,
        'Speaker 1': roles.speaker1,
        'Speaker 2': roles.speaker2,
        'Speaker 3': roles.speaker3,
        'Evaluator 1': roles.eval1,
        'Evaluator 2': roles.eval1,
        'Evaluator 3': roles.eval1,
        'Table Topics Master': roles.ttmaster,
        'Table Topics Evaluator': roles.tteval,
        'Timer': roles.timer,
        'Ah Counter': roles.ah_counter,
        'Ballot Counter': roles.facilitator,
        'Quizmaster': roles.quizmaster,
    }

    for old_key in rolelist:  # Iterate through the old rolelist
        for new_key in new_roles:  # Iterate through the new roles user has submitted
            if old_key == new_key:  # If the roles in both lists match (ex: Toastmaster and Toastmaster)
                if str(rolelist[old_key]) != str(new_roles[new_key]):  # If the new role assignment is different than the old role assignment
                    if old_key == 'Toastmaster':
                        roles.toastmaster = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Chair':
                        roles.chair = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Facilitator':
                        roles.facilitator = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Zoom Master':
                        roles.zoom = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'General Evaluator':
                        roles.geneval = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Speaker 1':
                        roles.speaker1 = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Speaker 2':
                        roles.speaker2 = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Speaker 3':
                        roles.speaker3 = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Evaluator 1':
                        roles.eval1 = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Evaluator 2':
                        roles.eval2 = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Evaluator 3':
                        roles.eval3 = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Table Topics Master':
                        roles.ttmaster = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Table Topics Evaluator':
                        roles.tteval = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Timer':
                        roles.timer = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Ah Counter':
                        roles.ah_counter = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Ballot Counter':
                        roles.facilitator = User.objects.filter(username=new_roles[new_key]).first()
                    elif old_key == 'Quizmaster':
                        roles.quizmaster = User.objects.filter(username=new_roles[new_key]).first()
                    
                    roles.save()  

    # Update the meeting eventlist
    eventlist = Eventlist.objects.filter(meeting=meeting).order_by('event_number').values()  # Get the existing events for the meeting
    print(new_roles)
    index = 1
    for event in eventlist:  # iterate through the existing eventlist and replace the entries
        try:
            event_changes = Eventlist.objects.get(id=event['id'])  # Get the event object to edit
            event_changes.description = new_roles[f'{index} description']
            event_changes.role = new_roles[f'{index} role']
            event_changes.duration = new_roles[f'{index} duration']
            event_changes.green_time = new_roles[f'{index} min_time']
            event_changes.yellow_time = new_roles[f'{index} mid_time']
            event_changes.red_time = new_roles[f'{index} max_time']

            event_changes.save()

            index += 1
        except:
            index += 1

    return


def update_one_role_in_database(role, rolelist, user):
    #Iterate through all the roles in plain text, if you find a match update the database entry
    if role == "Facilitator":
        if user == "None":
            rolelist.facilitator = None
        else:
            rolelist.facilitator = user
        rolelist.save()
        return True
    elif role == "Zoom Master":
        if user == "None":
            rolelist.zoom = None
        else:
            rolelist.zoom = user
        rolelist.save()
        return True
    elif role == "Toastmaster":
        if user == "None":
            rolelist.toastmaster = None
        else:
            rolelist.toastmaster = user
        rolelist.save()
        return True
    elif role == "General Evaluator":
        if user == "None":
            rolelist.geneval = None
        else:
            rolelist.geneval = user
        rolelist.save()
        return True
    elif role == "Chair":
        if user == "None":
            rolelist.chair = None
        else:
            rolelist.chair = user
        rolelist.save()
        return True
    
    elif role == "Speaker 1":
        if user == "None":
            rolelist.speaker1 = None
        else:
            rolelist.speaker1 = user
        rolelist.save()
        return True
    elif role == "Speaker 2":
        if user == "None":
            rolelist.speaker2 = None
        else:
            rolelist.speaker2 = user
        rolelist.save()
        return True
    elif role == "Speaker 3":
        if user == "None":
            rolelist.speaker3 = None
        else:
            rolelist.speaker3 = user
        rolelist.save()
        return True
    elif role == "Evaluator 1":
        if user == "None":
            rolelist.eval1 = None
        else:
            rolelist.eval1 = user
        rolelist.save()
        return True
    elif role == "Evaluator 2":
        if user == "None":
            rolelist.eval2 = None
        else:
            rolelist.eval2 = user
        rolelist.save()
        return True
    elif role == "Evaluator 3":
        if user == "None":
            rolelist.eval3 = None
        else:
            rolelist.eval3 = user
        rolelist.save()
        return True
    
    elif role == "Table Topics Master":
        if user == "None":
            rolelist.ttmaster = None
        else:
            rolelist.ttmaster = user
        rolelist.save()
        return True
    elif role == "Table Topics Evaluator":
        if user == "None":
            rolelist.tteval = None
        else:
            rolelist.tteval = user
        rolelist.save()
        return True
    
    elif role == "Timer":
        if user == "None":
            rolelist.timer = None
        else:
            rolelist.timer = user
        rolelist.save()
        return True
    elif role == "Ah Counter":
        if user == "None":
            rolelist.ah_counter = None
        else:
            rolelist.ah_counter = user
        rolelist.save()
        return True
    elif role == "Quizmaster":
        if user == "None":
            rolelist.quizmaster = None
        else:
            rolelist.quizmaster = user
        rolelist.save()
        return True

    return false


def meeting_roles_list(meeting):
    
    # Get the meeting information
    roles = Rolelist.objects.filter(meeting=meeting)

    # Create a list of all the roles and who has filled them
    rolelist = {
        'Toastmaster': roles[0].toastmaster,
        'Chair': roles[0].chair,
        'Facilitator': roles[0].facilitator,
        'Zoom Master': roles[0].zoom,
        'General Evaluator': roles[0].geneval,
        'Speaker 1': roles[0].speaker1,
        'Speaker 2': roles[0].speaker2,
        'Speaker 3': roles[0].speaker3,
        'Evaluator 1': roles[0].eval1,
        'Evaluator 2': roles[0].eval2,
        'Evaluator 3': roles[0].eval3,
        'Table Topics Master': roles[0].ttmaster,
        'Table Topics Evaluator': roles[0].tteval,
        'Timer': roles[0].timer,
        'Ah Counter': roles[0].ah_counter,
        'Ballot Counter': roles[0].zoom,
        'Quizmaster': roles[0].quizmaster,
    } 

    return rolelist


def create_default_eventlist(meeting):

    # List of all the default events
    # Format = [description, role, duration, green_time, yellow_time, red_time]
    default_eventlist = [
        ['Call Meeting to Order, Meeting Rules, Introduce Toastmasters, Welcome Guests', 'Chair', 3, 1, 2, 3],
        ['Introduce the Timer, Ah Counter, Quizmaster, and General Evaluator', 'Toastmaster', 5, 3, 4, 5],
        ['*Speech 1*\nPathway: \nLevel: \nProject: ', "Speaker 1", 7, 5, 6, 7],
        ['*Speech 2*\nPathway: \nLevel: \nProject: ', "Speaker 2", 7, 5, 6, 7],
        ['*Speech 3*\nPathway: \nLevel: \nProject: ', "Speaker 3", 7, 5, 6, 7],
        ['Best Speech Voting', "Zoom Master", 1, 0.5, 0.75, 1],
        ['Evaluation 1', "Evaluator 1", 3, 2, 2.5, 3],
        ['Evaluation 2', "Evaluator 2", 3, 2, 2.5, 3],
        ['Evaluation 3', "Evaluator 3", 3, 2, 2.5, 3],
        ['Table Topics', "Table Topics Master", 25, 15, 20, 25],
        ['Best Table Topics Speech Voting', "Zoom Master", 1, 0.5, 0.75, 1],
        ['Table Topics Evaluations', "Table Topics Evaluator", 4, 2, 3, 4],
        ["Timer's Report", "Timer", 3, 1, 2, 3],
        ["Ah Counter's Report", "Ah Counter", 3, 1, 2, 3],
        ["Quizmaster's Report", "Quizmaster", 3, 1, 2, 3],
        ["Ballot Counter's Report", "Zoom Master", 2, 1, 1, 2],
        ["General Evaluator's Report", "General Evaluator", 5, 3, 4, 5],
        ["Next Meeting Role/Speech Sign Ups", "Toastmaster", 3, 1, 2, 3],
        ["Club Business and Guest Feedback", "", 3, 1, 2, 3],
        ["Membership and Pathways Information", "", 3, 1, 2, 3],
        ["Group Photo", "Zoom Master", 1, 0.5, 0.75, 1],
        ["Meeting Adjourned", "", 0, 0, 0, 0]
    ]

    # Create each of the default events
    # Fill in the meeting properties for each meeting
    event_number = 1
    for event in default_eventlist:
        event = Eventlist(meeting=meeting, event_number=event_number, description=event[0], role=event[1], duration=event[2], green_time=event[3], yellow_time=event[4], red_time=event[5])
        event.save()
        event_number = event_number + 1

    # Get the eventlist for the meeting form the database and return it
    final_eventlist = Eventlist.objects.filter(meeting=meeting).order_by('event_number').values()

    return final_eventlist


def role_recommendation_list(meeting):
    recommendation_list = {'Toastmaster': [], 'Facilitator': [], 'Zoom Master': [],'Chair': [], 'General Evaluator': [], 'Speaker 1': [], 'Speaker 2': [], 'Speaker 3': [], 'Evaluator 1': [], 'Evaluator 2': [], 'Evaluator 3': [], 'Table Topics Master': [], 'Table Topics Evaluator': [], 'Timer': [], 'Ah Counter': [], 'Ballot Counter': [], 'Quizmaster': [], 'Sergeant At Arms': []}
    temp_rec = []

    # Get list of users with confirmed or unknown attendence
    attendees = Attendee.objects.filter(meeting=meeting).filter(Q(status="T") | Q(status="U"))

    # Get all user information
    users = []
    for attendee in attendees:
        user = {'user': attendee.user.username, 'speech_count': attendee.user.speech_count, 'meeting_count': attendee.user.meeting_count, 'other_role': "", 'days_since_last_time': ""}
        users.append(user)

    # Sort users by experience level
    users.sort(key=sortByMeetingCountFunc)

    # Iterate through each of the meeting roles and get a recommendation list for each
    for role in recommendation_list:

        # Add all the users who will be attending
        for user in users:
            recommendation_list[role].append(user['user'])

        # Use the past role holders function to get the list of people who have held the role in the past time period
        past_holders = past_role_holders(meeting, role)

        for index, recommendation in enumerate(recommendation_list[role]):
            if past_holders == False:
                continue

            for user_info in past_holders:
                if recommendation == user_info['user']:
                    del recommendation_list[role][index]
                    recommendation_list[role].append(user_info)

        # Figure out and label the users who already have a role this meeting
        rolelist = meeting_roles_list(meeting)

        # Iterate through the existing list of recommendations for the role
        for index, recommendation in enumerate(recommendation_list[role]):

            # Iterate throguh the rolelist for the selected meeting
            for meeting_role in rolelist:

                # Check if the user being recommended already holds a role this meeting
                if str(recommendation) == str(rolelist[meeting_role]):

                    # Add the role to the user string to display in the dropdown role selector menu
                    recommendation_list[role][index] = " ".join((recommendation_list[role][index], "[" + str(meeting_role) + "]"))
                
                # Check if the user is a past role holder, they have to be treated differently because they use a dictionary instead of a plain string
                for key in recommendation:
                    if key == 'user':

                        # Check if the user being recommended already holds a role this meeting
                        if str(recommendation[key]) == str(rolelist[meeting_role]):

                            # Add the role to the user string to display in the dropdown role selector menu
                            recommendation_list[role][index]['user'] = " ".join((recommendation_list[role][index]['user'], "[" + str(meeting_role) + "]"))

    return recommendation_list


def sortByMeetingCountFunc(e):
    return e['meeting_count']

def sortByTimeSinceFunc(e):
    return e['time_since']


def past_role_holders(meeting, role):

    # Get the past 2 months of meetings
    startdate = meeting.starttime - timedelta(days=62)
    enddate = meeting.starttime - timedelta(days=1)
    meeting_list_query = Meeting.objects.filter(starttime__range=[startdate, enddate])

    # Check if there are past meetings
    if not meeting_list_query:
        return False

    # Get the role holder for each meeting
    recommendations = []
    for past_meeting in meeting_list_query:
        
        # Get who held the role at the meeting
        role_holder = ""
        role_list = Rolelist.objects.get(meeting=past_meeting)
        for field in role_list._meta.fields:
            if field.name == convert_role_to_shorthand(role):
                role_holder = str(getattr(role_list, field.name))

        # Get the time since the meeting
        time_since_role = meeting.starttime - past_meeting.starttime
        time_since_role = time_since_role.days

        # Assign a color to indicate how long it has been
        if time_since_role < 14:
            indicator = 'red'
        elif time_since_role < 31:
            indicator = '#CCCC00'
        else:
            indicator = 'green'
        
        recommendations.append({'user': role_holder, 'time_since': time_since_role, 'indicator': indicator})

    return recommendations


# Function to convert the plain text display versions of roles to the shorthand I use for the database and backend
def convert_role_to_shorthand(role):
    #Iterate through all the roles, if you find a match return the shorthand version of the role
    if role == "Facilitator":
        return "facilitator"
    elif role == "Zoom Master":
        return "zoom"
    elif role == "Toastmaster":
        return "toastmaster"
    elif role == "General Evaluator":
        return "geneval"
    elif role == "Chair":
        return "chair"
    
    elif role == "Speaker 1":
        return "speaker1"
    elif role == "Speaker 2":
        return "speaker2"
    elif role == "Speaker 3":
        return "speaker3"
    elif role == "Evaluator 1":
        return "eval1"
    elif role == "Evaluator 2":
        return "eval2"
    elif role == "Evaluator 3":
        return "eval3"
    
    elif role == "Table Topics Master":
        return "ttmaster"
    elif role == "Table Topics Evaluator":
        return "tteval"
    
    elif role == "Timer":
        return "timer"
    elif role == "Ah Counter":
        return "ah_counter"
    elif role == "Quizmaster":
        return "quizmaster"
    elif role == "Ballot Counter":
        return "facilitator"

    return "Invalid Role"